File: pay.py
# Cenário
#Vamos supor que você tem uma aplicação que precisa lidar com uma interface de pagamento.
#A aplicação espera uma interface com métodos como process_payment(amount), mas você tem
#uma classe existente que só oferece uma interface com um método diferente, make_transaction(amount).

#Vamos criar um adaptador para converter a chamada de process_payment(amount) para make_transaction(amount).



# Interface esperada pela aplicação
class PaymentProcessor:
    def process_payment(self, amount):
        raise NotImplementedError("Subclass must implement this method")

# Classe com a interface existente
class OldPaymentService:
    def make_transaction(self, amount):
        print(f"Processing transaction of ${amount}")

# Adaptador que adapta OldPaymentService para PaymentProcessor
class PaymentAdapter(PaymentProcessor):
    def __init__(self, old_payment_service):
        self.old_payment_service = old_payment_service

    def process_payment(self, amount):
        self.old_payment_service.make_transaction(amount)

# Uso do adaptador
def main():
    old_service = OldPaymentService()
    adapter = PaymentAdapter(old_service)
    
    # A aplicação usa a interface PaymentProcessor
    adapter.process_payment(100)

if __name__ == "__main__":
    main()

File: pay.java
public class AdapterPatternExample {

    // Interface PagamentoProcessor
    interface PagamentoProcessor {
        void processarPagamento(double quantia);
    }

    // Implementação existente com PayPalProcessor
    static class PayPalProcessor implements PagamentoProcessor {
        @Override
        public void processarPagamento(double quantia) {
            System.out.println("Pagamento de " + quantia + " processado via PayPal.");
        }
    }

    // Nova classe Stripe com interface diferente
    static class Stripe {
        public void makePayment(double amount) {
            System.out.println("Pagamento de " + amount + " processado via Stripe.");
        }
    }

    // Classe StripeAdapter que implementa PagamentoProcessor
    static class StripeAdapter implements PagamentoProcessor {
        private Stripe stripe;

        public StripeAdapter(Stripe stripe) {
            this.stripe = stripe;
        }

        @Override
        public void processarPagamento(double quantia) {
            stripe.makePayment(quantia); 
        }
    }

    // Método main para testar o padrão Adapter
    public static void main(String[] args) {
        // Usando o processador PayPal
        PagamentoProcessor paypal = new PayPalProcessor();
        paypal.processarPagamento(50.0); // Saída: "Pagamento de 50.0 processado via PayPal."

        // Usando o processador Stripe através do adaptador
        Stripe stripe = new Stripe();
        PagamentoProcessor stripeAdapter = new StripeAdapter(stripe);
        stripeAdapter.processarPagamento(75.0); // Saída: "Pagamento de 75.0 processado via Stripe."
    }
}
